✨ Use bulk_write for save operations, including referenced documents

tiangolo opened the pull request:
✨ Use bulk_write for save operations, including referenced documents

art049 commented on odmantic/engine.py:
```diff
@@ -291,6 +293,39 @@ def _prepare_find_pipeline(
+        *,
+        collection_updates: Union[CollectionUpdatesType, None] = None,
+    ) -> CollectionUpdatesType:
+        """Perform an atomic save operation in the specified session"""
```
This would be a bit clearer:
```suggestion
        """Generate update operations"""
```

tiangolo replied:
Right! Copy-paste mistake 🙈

art049 commented on odmantic/engine.py:
```diff
@@ -511,28 +551,37 @@ async def find_one(
+        # saved first
+        for collection_name, updates in reversed(list(collection_updates.items())):
+            collection = self._get_collection_from_name(collection_name)
+            update_operations = [update[0] for update in updates]
+            update_instances = [update[1] for update in updates]
```
`[update[0] for update in updates]` is a bit unusual to see for me.

Maybe this would be a bit clearer
```python
update_operations, update_instances = zip(*updates)
```

But still quite strange, maybe the easiest way would be to build a small dataclass or a namedtuple to have something a bit more explicit.

tiangolo replied:
Sure thing! Done ✔️

art049 commented on odmantic/engine.py:
```diff
@@ -511,28 +551,37 @@ async def find_one(
+            if e.details["writeErrors"][0]["code"] == 11000:
                 raise DuplicateKeyError(instance, e)
-            object.__setattr__(instance, "__fields_modified__", set())
+            raise  # pragma: no cover
```
Maybe we could try to reach this, WDYT ?

tiangolo replied:
I have been trying to trigger an error that shows up here for the past hours but I haven't figured out something that would error out here and that we can trigger in tests without monkeypatching everything (that would defeat the purpose of the test).

But maybe there's a simple thing we could test that would certainly raise here and I just don't know it yet.

If you have any ideas, let me know!

tiangolo replied:
I have given up about reaching this `raise` 😔 

I've been two days trying to find something that would trigger a `BulkWriteError` that is not a `DuplicateKeyError` and that we can trigger from some simple update in ODMantic models. I haven't found anything, there might not be any other type of error, but there's no way to be sure, and if we don't raise here, we would end up swallowing any other potential or future exception there.

But I can't find something that would raise something that goes to this code path, I tried a lot of things and got a lot of different exceptions, but nothing inside of this code path.

I even tried looking a bit at the source code of MongoDB itself, looking for those keys, but there's nothing obvious that gives away what other types of errors could end up there.

If you have any ideas, please let me know! (or feel free to take over and add commits).

art049 replied:
Long time :upside_down_face: . It seems hard to reach this line.

art049 commented on odmantic/engine.py:
```diff
@@ -511,28 +551,37 @@ async def find_one(
+                collection_updates=collection_updates, session=session
+            )
+        except pymongo.errors.BulkWriteError as e:
+            if e.details["writeErrors"][0]["code"] == 11000:
```
Maybe we could add a reference to the error code list in a comment.
The only thing I found is this https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml
But i think it would be helpful if something goes wrong.

tiangolo replied:
Cool! I added it. Indeed, there are practically no docs about the possible errors and when they are thrown. 😢

art049 commented on odmantic/exceptions.py:
```diff
@@ -43,7 +44,11 @@ class DuplicateKeyError(BaseEngineException):
-        self, instance: "Model", driver_error: pymongo.errors.DuplicateKeyError
+        self,
+        instance: "Model",
+        driver_error: Union[
+            pymongo.errors.DuplicateKeyError, pymongo.errors.BulkWriteError
+        ],
```
I think now we only need `BulkWriteError` since we're not using update directly anymore.
```suggestion
        driver_error: pymongo.errors.BulkWriteError,
```

tiangolo replied:
Yep, makes sense! 🚀

art049 reviewed (commented):
Just a few details to fix but awesome work 🔥.
Also, really nice `_get_collection_from_name` refactor 😉

tiangolo commented:
Thanks for the review! I added the changes. 🤓 

There's only one that I don't really know how to do, I added a comment about it.

I also included/merged https://github.com/art049/odmantic/pull/308 in this PR to be able to run CI and check if I'm breaking anything else.

tiangolo commented:
I'm struggling to debug the current errors, I had never used tox outside this project. I managed to install and setup all the dependencies needed by tox (a python installation for each version, with the exact name, etc). And I was able to replicate the error running a single tox environment with:

```
$ tox -e python3.9-motor24-pydantic18
```

And it consistently errors out the same way as in CI.

Then I'm trying to replicate it outside of tox to be able to investigate and debug it, I installed a local python3.9, motor 2.4, pydantic 1.8, and ran the same test, and it always passed.

I'm not sure what else to try as it seems like something particular to the tox-specific environment that I can't access and debug. Or maybe I'm doing something wrong, but I ran out of ideas of what else to try to check it. 🤔  ...also to see if it's an actual new error or some quirk of the setup I have.

art049 commented:
I think we can probably merge after the v1 release :)
Btw with the bench, the speedup is amazing; this is super exciting, thank you. :clap:
